In [ ]:
#Wrangling Assignment - Avery Anderson kdu7bp

In [1]:
import numpy as np # Import the numpy package into your workspace
import matplotlib.pyplot as plt # Import matplotlib into your workspace
import pandas as pd  # Import the pandas package into your workspace

In [3]:
df = pd.read_csv('/Users/averyanderson/Desktop/data/airbnb_hw.csv',low_memory=False)

In [5]:
print(df.shape, '\n') # List the dimensions of df
print(df.dtypes, '\n') # The types of the variables
print(df.columns, '\n') # Column names

(30478, 13) 

Host Id                         int64
Host Since                     object
Name                           object
Neighbourhood                  object
Property Type                  object
Review Scores Rating (bin)    float64
Room Type                      object
Zipcode                       float64
Beds                          float64
Number of Records               int64
Number Of Reviews               int64
Price                          object
Review Scores Rating          float64
dtype: object 

Index(['Host Id', 'Host Since', 'Name', 'Neighbourhood ', 'Property Type',
       'Review Scores Rating (bin)', 'Room Type', 'Zipcode', 'Beds',
       'Number of Records', 'Number Of Reviews', 'Price',
       'Review Scores Rating'],
      dtype='object') 



In [7]:
df.head(5)

Host Id Host Since                                Name Neighbourhood   \
0   5162530        NaN     1 Bedroom in Prime Williamsburg       Brooklyn   
1  33134899        NaN     Sunny, Private room in Bushwick       Brooklyn   
2  39608626        NaN                Sunny Room in Harlem      Manhattan   
3       500  6/26/2008  Gorgeous 1 BR with Private Balcony      Manhattan   
4       500  6/26/2008            Trendy Times Square Loft      Manhattan   

  Property Type  Review Scores Rating (bin)        Room Type  Zipcode  Beds  \
0     Apartment                         NaN  Entire home/apt  11249.0   1.0   
1     Apartment                         NaN     Private room  11206.0   1.0   
2     Apartment                         NaN     Private room  10032.0   1.0   
3     Apartment                         NaN  Entire home/apt  10024.0   3.0   
4     Apartment                        95.0     Private room  10036.0   3.0   

   Number of Records  Number Of Reviews Price  Review Scores Rating  
0                  1                  0   145                   NaN  
1                  1                  1    37                   NaN  
2                  1                  1    28                   NaN  
3                  1                  0   199                   NaN  
4                  1                 39   549                  96.0

In [9]:
df['Price'].unique()

array(['145', '37', '28', '199', '549', '149', '250', '90', '270', '290',
       '170', '59', '49', '68', '285', '75', '100', '150', '700', '125',
       '175', '40', '89', '95', '99', '499', '120', '79', '110', '180',
       '143', '230', '350', '135', '85', '60', '70', '55', '44', '200',
       '165', '115', '74', '84', '129', '50', '185', '80', '190', '140',
       '45', '65', '225', '600', '109', '1,990', '73', '240', '72', '105',
       '155', '160', '42', '132', '117', '295', '280', '159', '107', '69',
       '239', '220', '399', '130', '375', '585', '275', '139', '260',
       '35', '133', '300', '289', '179', '98', '195', '29', '27', '39',
       '249', '192', '142', '169', '1,000', '131', '138', '113', '122',
       '329', '101', '475', '238', '272', '308', '126', '235', '315',
       '248', '128', '56', '207', '450', '215', '210', '385', '445',
       '136', '247', '118', '77', '76', '92', '198', '205', '299', '222',
       '245', '104', '153', '349', '114', '320', '292', '22

In [11]:
df['Price'] = pd.to_numeric(df['Price'], errors='coerce') # Coerce the variable to numeric
df['price_nan'] = df['Price'].isnull() # Equals 1 if missing, 0 if non-null
print('After coercion: \n', df['Price'].describe(),'\n') # Describe the numeric variable
print('Total Missings: \n', sum(df['price_nan']),'\n') # How many missing values are there?

After coercion: 
 count    30297.000000
mean       153.357032
std        108.895698
min         10.000000
25%         80.000000
50%        125.000000
75%        192.000000
max        999.000000
Name: Price, dtype: float64 

Total Missings: 
 181 



^^ Above I realized that the values with commas could not be used as numerical integers. Using the code learned in the wrangling lesson in class, I coerced the column to be all numerical which resulted in 118 missing values

Q1 Part B: Minnesota Data

In [15]:
df2 = pd.read_csv('/Users/averyanderson/Desktop/data/mn_police_use_of_force.csv',low_memory=False)

In [51]:
df2.head(20)

response_datetime               problem is_911_call primary_offense  \
0   2016/01/01 00:47:36  Assault in Progress          Yes          DASLT1   
1   2016/01/01 02:19:34                Fight           No          DISCON   
2   2016/01/01 02:19:34                Fight           No          DISCON   
3   2016/01/01 02:28:48                Fight           No          PRIORI   
4   2016/01/01 02:28:48                Fight           No          PRIORI   
5   2016/01/01 02:28:48                Fight           No          PRIORI   
6   2016/01/01 02:28:48                Fight           No          PRIORI   
7   2016/01/01 02:28:48                Fight           No          PRIORI   
8   2016/01/01 02:28:48                Fight           No          PRIORI   
9   2016/01/01 02:38:23                Fight           No          DISCON   
10  2016/01/01 02:38:23                Fight           No          DISCON   
11  2016/01/01 02:39:47    Suspicious Person           No          DISCON   
12  2016/01/01 02:39:47    Suspicious Person           No          DISCON   
13  2016/01/01 02:39:47    Suspicious Person           No          DISCON   
14  2016/01/01 03:11:20        Chase on Foot           No            FLEE   
15  2016/01/01 03:11:20        Chase on Foot           No            FLEE   
16  2016/01/01 03:22:39                Fight           No           FORCE   
17  2016/01/01 03:22:39                Fight           No           FORCE   
18  2016/01/01 15:22:04    Check the Welfare          Yes          THREAT   
19  2016/01/01 17:47:46      Unwanted Person          Yes           FORCE   

   subject_injury         force_type   force_type_action             race  \
0             NaN       Bodily Force  Body Weight to Pin            Black   
1             NaN  Chemical Irritant       Personal Mace            Black   
2             NaN  Chemical Irritant       Personal Mace            White   
3             NaN  Chemical Irritant  Crowd Control Mace            Black   
4             NaN  Chemical Irritant  Crowd Control Mace            Black   
5             NaN  Chemical Irritant  Crowd Control Mace            Black   
6             NaN  Chemical Irritant  Crowd Control Mace            Black   
7             NaN  Chemical Irritant  Crowd Control Mace            Black   
8             NaN  Chemical Irritant  Crowd Control Mace            Black   
9              No  Chemical Irritant  Crowd Control Mace            Black   
10             No  Chemical Irritant  Chemical Munitions            Black   
11            Yes       Bodily Force  Body Weight to Pin            Black   
12            Yes       Bodily Force          Joint Lock            Black   
13            Yes       Bodily Force  Body Weight to Pin            Black   
14             No       Bodily Force  Body Weight to Pin            Black   
15             No       Bodily Force               Knees            Black   
16            NaN  Chemical Irritant       Personal Mace            Black   
17            NaN  Chemical Irritant       Personal Mace            Black   
18            NaN       Bodily Force  Body Weight to Pin  Native American   
19             No       Bodily Force               Knees            White   

       sex   age        type_resistance  precinct     neighborhood  
0     Male  20.0                 Tensed         1    Downtown East  
1   Female  27.0  Verbal Non-Compliance         1    Downtown West  
2   Female  23.0  Verbal Non-Compliance         1    Downtown West  
3     Male  20.0    Commission of Crime         1    Downtown West  
4     Male  20.0    Commission of Crime         1    Downtown West  
5     Male  20.0    Commission of Crime         1    Downtown West  
6   Female  20.0    Commission of Crime         1    Downtown West  
7     Male  20.0    Commission of Crime         1    Downtown West  
8     Male  20.0            Unspecified         1    Downtown West  
9     Male  18.0    commission of crime         1    Downtown West  
10    Male 

In [31]:
df2['subject_injury'].unique()

array([nan, 'No', 'Yes'], dtype=object)

In [33]:
total = len(df2)
print(total)

12925


In [45]:
print(df2['subject_injury'].unique(), '\n') 
df2['subject_injury'] = df['subject_injury'].replace(' ',np.nan) 
print(df2['subject_injury'].value_counts(), '\n')

[nan 'No' 'Yes'] 

subject_injury
Yes    1631
No     1446
Name: count, dtype: int64 



Of 12,925 total datapoints, only 3,077 are a yes or a no which is concerning because it means that a majority of this information is missing. Only 23.8% have the yes or no

In [53]:
cross_tab = pd.crosstab(df2['force_type'], df2['subject_injury'])
print(cross_tab)

subject_injury            No   Yes
force_type                        
Baton                      0     2
Bodily Force            1093  1286
Chemical Irritant        131    41
Firearm                    2     0
Gun Point Display         33    44
Improvised Weapon         34    40
Less Lethal Projectile     1     2
Police K9 Bite             2    44
Taser                    150   172


It appears that most of the data known about injuries comes from instances with bodily force, especially compared to some other cases like those of chemical irritants

Q1-C: 

In [61]:
df3 = pd.read_parquet('/Users/averyanderson/Desktop/data/justice_data.parquet') #reading in data

In [63]:
df3.head(10)

InternalStudyID REQ_REC# Defendant_Sex Defendant_Race Defendant_BirthYear  \
0        ADI00001        1             M              W                1986   
1        ADI00007        3             M              B                1956   
2        ADI00008        4             M              W                1990   
3        CDI00036        6             M              B                1989   
4        CDI00038        7             F              W                1988   
5        CDI00039        8             M              B                1978   
6        CDI00041        9             F              B                1981   
7        CDI00043       10             M              B                1999   
8        CDI00044       11             M              B                1990   
9        CDI00046       12             M              W                1996   

  Defendant_Age  Defendant_AgeGroup Defendant_AgeatCurrentArrest  \
0            31                   3                           31   
1            60                   6                           60   
2            27                   3                           27   
3            27                   3                           27   
4            28                   3                           28   
5            39                   4                           39   
6            35                   3                           35   
7            18                   2                           18   
8            27                   3                           27   
9            21                   2                           21   

   Defendant_AttorneyTypeAtCaseClosure  Defendant_IndigencyStatus  ...  \
0                                   99                         99  ...   
1                                    9                          9  ...   
2                                    9                          9  ...   
3                                    0                          0  ...   
4                                    0                          0  ...   
5                                    0                          0  ...   
6                                    1                          1  ...   
7                                    0                          0  ...   
8                                    1                          1  ...   
9                                    1                          1  ...   

  NewFelonySexualAssaultArrest_OffDate  \
0                                        
1                                        
2                                        
3                                        
4                                        
5                                        
6                                        
7                                        
8                                        
9                                        

  NewFelonySexualAssaultArrest_ArrestDate  \
0                                           
1                                           
2                                           
3                                           
4                                           
5                                           
6                                           
7                                           
8                                           
9                                           

   NewFelonySexualAssaultArrest_DaysBetweenContactEventandOffDate  \
0                                                                   
1                                                                   
2                                                                   
3                                                                   
4                                                                   
5                                                                   
6                                                                   
7                                                        

In [65]:
var = 'WhetherDefendantWasReleasedPretrial' # A Dummy Example
print(df3[var].unique(), '\n')
df3[var] = df3[var].replace( " ", np.nan) 
print(df3[var].value_counts()) 

[9 0 1] 

WhetherDefendantWasReleasedPretrial
1    19154
0     3801
9       31
Name: count, dtype: int64


Q1 - D: Cleaning ImposedSentenceAllChargeInContactEvent

In [71]:
df3['ImposedSentenceAllChargeInContactEvent'].head(10) #first look at the column

0                    
1                  60
2                  12
3    .985626283367556
4                    
5                  12
6                  36
7                   6
8                  24
9                  12
Name: ImposedSentenceAllChargeInContactEvent, dtype: object

In [73]:
df3['SentenceTypeAllChargesAtConvictionInContactEvent'].head(10) #looking at this column for comparison

0    9
1    0
2    1
3    1
4    4
5    0
6    1
7    0
8    1
9    1
Name: SentenceTypeAllChargesAtConvictionInContactEvent, dtype: int64

In [77]:
df3['ImposedSentenceAllChargeInContactEvent'] = pd.to_numeric(df3['ImposedSentenceAllChargeInContactEvent'], errors='coerce') # Coerce the variable to numeric
df3['sentence_x'] = df3['ImposedSentenceAllChargeInContactEvent'].isnull() # Equals 1 if missing, 0 if non-null
print('After coercion: \n', df3['ImposedSentenceAllChargeInContactEvent'].describe(),'\n') # Describe the numeric variable

After coercion: 
 count    13933.000000
mean        19.899671
std         63.571627
min          0.000000
25%          0.000000
50%          2.000000
75%         12.000000
max       2208.000000
Name: ImposedSentenceAllChargeInContactEvent, dtype: float64 



I noticed that the conviction sentence column was all integers, which pointed me towards changing the imposed sentence type from object to a float since these datapoints are decimals

Question 2: Sharks

In [83]:
df4 = pd.read_excel('/Users/averyanderson/Desktop/data/GSAF5.xlsx')

In [85]:
df4.head(10)

Date    Year        Type        Country  \
0  2025-01-23 00:00:00  2025.0  Unprovoked      Australia   
1  2025-01-22 00:00:00  2025.0  Unprovoked      Australia   
2  2025-01-11 00:00:00  2025.0    Provoked            USA   
3  2025-01-02 00:00:00  2025.0  Unprovoked  New Caledonia   
4  2025-01-02 00:00:00  2025.0  Unprovoked      Australia   
5  2024-12-29 00:00:00  2024.0  Unprovoked          Egypt   
6  2024-12-29 00:00:00  2024.0  Unprovoked          Egypt   
7  2024-12-28 00:00:00  2024.0  Unprovoked      Australia   
8  2024-12-02 00:00:00  2024.0  Unprovoked      Australia   
9  2024-12-01 00:00:00  2024.0  Unprovoked      Australia   

                 State                                           Location  \
0    Western Australia                                     Shipwreck Cove   
1    Western Australia                                     Cheynes beach    
2               Hawaii                      Off Haleiwa Boat Harbour Oahu   
3         Grande Terre                        Islet of Kendek near Koumac   
4      South Australia            Granites Beach near Westall Streaky Bay   
5  North of Marsa Alam                                            Red Sea   
6  North of Marsa Alam                                            Red Sea   
7           Queensland                   Humpy Island Great Keppel Island   
8           Queensland     Curtis Island Near Gladstone QLD central coast   
9    Western Australia  Greenough River Mouth Beach Cape Burney 12km s...   

       Activity                                    Name Sex  Age  ...  \
0       Bathing                                    Boy    M   12  ...   
1       Surfing                             Dale Kittow   M   37  ...   
2        Diving  Male not stated was a dive tour worker   M   23  ...   
3  Spearfishing                        Robert Cuewapuru   M   40  ...   
4       Surfing                           Lance Appleby   M   28  ...   
5  SCUBA Diving                         Peppino Fappani   M   69  ...   
6  SCUBA Diving                       Gianluca Di Gioia   M   48  ...   
7  Spearfishing                            Luke Walford   M   40  ...   
8  Spearfishing                              Not stated   M  60+  ...   
9       Surfing                  James Sweetman (child)   M    ?  ...   

                                 Species   \
0                              Not stated   
1                           Bronze whaler   
2                              Not stated   
3          Reportedly Tiger or Bull shark   
4  Great White Shart estimated 4.8m 16 ft   
5                Reportedly a Tiger Shark   
6                Reportedly a Tiger Shark   
7                                Unknown    
8                              Not stated   
9                1m shark species unknown   

                                              Source  pdf href formula href  \
0                                     Bob Myatt GSAF  NaN          NaN  NaN   
1                                     Bob Myatt GSAF  NaN          NaN  NaN   
2                  Kevin McMurray Trackingsharks.com  NaN          NaN  NaN   
3                      Johannes Marchand Todd Smith   NaN          NaN  NaN   
4  Glen Folkard: Simon De Marchi News.com.au: The...  NaN          NaN  NaN   
5  Todd Smith : Kevin McMurray Trackingsharks .co...  NaN          NaN  NaN   
6  Todd Smith : Kevin McMurray Trackingsharks .co...  NaN          NaN  NaN   
7  Todd Smith Glen Folkard ABC News: The North We...  NaN          NaN  NaN   
8                                  Glen Folkard GSAF  NaN          NaN  NaN   
9                                     Bob Myatt GSAF  NaN          NaN  NaN   

  Case Number Case Number.1 original order Unnamed: 21 Unnamed: 22  
0         NaN           NaN            NaN         NaN         NaN  
1         NaN           NaN            NaN         NaN         NaN  
2         NaN           NaN            NaN         NaN         NaN  
3         NaN           NaN            NaN         N

In [91]:
df4['Year'].max()

2026.0

In [93]:
df4['Year'].min()

0.0

Question 4 Census Data Responses:
1. The 2020 Census collected data on race by allowing write in answers for the first time. Expanding from previous forms of the question, the 2020 census allowed those filling it out to check a box and then further specify race by writing it in. There was also an "other" category which allowed individuals to write in a race that was not previously listed.
2. The Census data is one of the most comprehensive sets of recent data of the US population and that data has real implications. The Census data is used to allocate federal funding and influence policy decisions. In addition to the uses by the government, other researchers will further use this data in their work. This data matters because it has real impacts on communities as people look at this data to determine decisions from funding to potential development projects.
3. The 2020 Census took a big step towards being more inclusive by enabling write in options. However, when dealing with data of real people, collecting that data with care is important. The current Census model still does not have a clear way for Americans with mixed race backgrounds to answer that question except by checking multiple boxes, which leaves unknowns. The balance between specific, personal data and data that is actually feasible to work with is a fine line. The more personal one allows the data to be, the harder it is to work with. But, dealing with real individuals and their background and seeking to accurately represent the US population, having those specifics is important. The write in option may complicate the data cleaning process, but it also allows more people to feel like they have a way to accurately represent themselves.
4. The 2020 Census only asked a question on biological sex with two answers being male or female. This question differs from the direction the Census was heading with the question about Race, which allowed much more flexibility in terms of what people could answer. Sex and gender have widely been considered two different entities and categories through which people identify. By limiting the answer choices to the question on sex without having a question on gender, the Census limits it accuracy in data. This data is used to examine issues like the gender wage gap, but by limiting the questions to sex, the Census looses all ability to analyze data from both transgender and non-binary individuals. If the census truly seeks to be able to analyze data on all US citizens, its limited answer choices mean that it is failing to represent a portion of those people.
5. Individual data is more complicated. It doesn't fit into boxes easy to categorize which can tempt those cleaning it to want to remove or simplify it, defeating the purpose of it in the first place. It would be a lost easier to analyze data on race if there are only 5 possible responses, but that would not be representative of the population. That data would be easier, but it would be false. The bad practice of deleting complicated data could be tempting. However, each day new technology in data analysis is making it easier to examine more complex data. Maybe AI can be a tool to parse through more options of responses. Just as there is potential for bad practices, there is also potential for good as people work to get more accurate and more representative data.
6. As I have mentioned before, when dealing with data of people, simplification can always be worrying. The fear that reducing someone's identity to a code or number can present problems in generalizations. Creating an algorithm to determine those categories would also require clear definitions of those categories. Who is responsible for designing an algorithm that defines what categories of gender or race there are? While there is potential for a better system that could help make accurate but easily analyzed data, these issues are ones that need to be considered. 